language/move-vm/runtime/src/loader.rs

```diff
@@ -0,0 +1,1378 @@
+                Type::Vector(Box::new(inner_type))
+            }
+            SignatureToken::Reference(inner_tok) => {
+                let inner_type = self.make_type(binary, inner_tok)?;
+                Type::Reference(Box::new(inner_type))
+            }
+            SignatureToken::MutableReference(inner_tok) => {
+                let inner_type = self.make_type(binary, inner_tok)?;
+                Type::MutableReference(Box::new(inner_type))
+            }
+            SignatureToken::Struct(sh_idx) => {
+                let struct_handle = binary.struct_handle_at(*sh_idx);
+                let struct_name = binary.identifier_at(struct_handle.name);
+                let module_handle = binary.module_handle_at(struct_handle.module);
+                let module_id = ModuleId::new(
+                    *binary.address_identifier_at(module_handle.address),
+                    binary.identifier_at(module_handle.name).to_owned(),
+                );
+                let def_idx = self.find_struct_by_name(struct_name, &module_id)?.0;
+                Type::Struct(def_idx)
+            }
+            SignatureToken::StructInstantiation(sh_idx, tys) => {
+                let type_parameters: Vec<_> = tys
+                    .iter()
+                    .map(|tok| self.make_type(binary, tok))
+                    .collect::<VMResult<_>>()?;
+                let struct_handle = binary.struct_handle_at(*sh_idx);
+                let struct_name = binary.identifier_at(struct_handle.name);
+                let module_handle = binary.module_handle_at(struct_handle.module);
+                let module_id = ModuleId::new(
+                    *binary.address_identifier_at(module_handle.address),
+                    binary.identifier_at(module_handle.name).to_owned(),
+                );
+                let def_idx = self.find_struct_by_name(struct_name, &module_id)?.0;
+                Type::StructInstantiation(def_idx, type_parameters)
+            }
+        };
+        Ok(res)
+    }
+
+    fn find_struct_by_name(
+        &self,
+        struct_name: &IdentStr,
+        module_id: &ModuleId,
+    ) -> VMResult<(usize, Arc<StructType>)> {
+        for (idx, ty) in self.structs.iter().enumerate() {
+            if struct_match(ty, &module_id, struct_name) {
+                return Ok((idx, Arc::clone(ty)));
+            }
+        }
+        Err(VMStatus::new(StatusCode::UNREACHABLE)
+            .with_message("Missing type in cache".to_string()))
+    }
+
+    fn resolve_function_handle(
+        &self,
+        func_name: &IdentStr,
+        module_id: &ModuleId,
+    ) -> VMResult<usize> {
+        for (idx, f) in self.functions.iter().enumerate() {
+            if function_match(&f, module_id, func_name) {
+                return Ok(idx);
+            }
+        }
+        Err(VMStatus::new(StatusCode::UNREACHABLE)
+            .with_message("Missing function in cache".to_string()))
+    }
```

Comment:
Use `BinaryCache<(ModuleId, IdentStr), Arc<_>>` instead of Vec<Arc<_>> for structs/functions in ModuleCache? `find_struct_by_name` and `resolve_function_handle` are called whenever we resolves an external symbol so it would be great if we can do better than an O(n) lookup.

Reply:
yes I know and glad you noticed, I also mentioned that above in the comment to the `BinaryCache`. 
I am not sure yet what the impact is given the number of types and functions in V1 and went down this path for now. I want this to go in like this and review later. Definitely on our list of things to review and play with.
That is also the part I mentioned in the overall comment about performance work that is "easy" to do and we can plug in later